File: JavaScript/1235-Maximum Profit in Job Scheduling.js
/**
 * @param {number[]} startTime
 * @param {number[]} endTime
 * @param {number[]} profit
 * @return {number}
 */
var jobScheduling = function(startTime, endTime, profit) {
    const N = startTime.length;
    let jobs = new Array();
    for (let i=0; i<N; i++){
        jobs.push([startTime[i], endTime[i], profit[i]]);
    }
    jobs.sort(function(a,b){
        return a[0] - b[0];
    })
    
    let getNext = (i) =>{
        if (i === N-1) return N;
        let left = i+1, right = N-1;
        while (left+1<right){
            let mid = left + ((right-left)>>1);
            if (jobs[mid][0]>=jobs[i][1]){
                right = mid;
            }else{
                left = mid;
            }
        }
        if (jobs[left][0] >= jobs[i][1]){
            return left;
        }else if(jobs[right][0] >= jobs[i][1]){
            return right;
        }else{
            return N;
        }
    }
    
    let memo = new Map();
    let dfs = (index) =>{
        if (index === N) return 0;
        if (memo.has(index)){
            return memo.get(index);
        }
        let res = 0, nextIdx = getNext(index);
        res = Math.max(jobs[index][2] + dfs(nextIdx), dfs(index+1) );
        memo.set(index, res);
        return res;
    }
    
    return dfs(0);
};

File: C++/1235-Maximum Profit in Job Scheduling.cpp
class Solution {
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        vector<pair<int, int>> arr;
        for (int i=0; i<startTime.size(); i++){
            arr.push_back(make_pair(endTime[i], i));
        }
        
        sort(arr.begin(), arr.end());

        map<int, int> preMap {{0, 0}};
        int preVal = 0;
        
        for (auto v:arr) {
            int end = v.first;
            int index = v.second;
            int profits = profit[index];
            int start = startTime[index];
            
            auto it = preMap.upper_bound(start);
            it = prev(it);
            // if (it != preMap.begin()) it = prev(it);
            
            int currVal = preMap[it->first] + profits;
            
            if (currVal > preVal){
                preMap[end] = currVal;
                preVal = currVal;
            }
        }
        return preMap.rbegin()->second;
    }
};